### ISSUE TITLE
Error: version `GLIBC_2.29' not found, Failed to serve assembly starter project using Docker Image - 

### ISSUE THREAD

**Issue by byteshiva** [fastly/cli#513]
Failed to locally serve using viceroy for `Dockerfile-assemblyscript` instance

Refer:https://github.com/fastly/cli/blob/main/Dockerfile-assemblyscript

Error:
```
/root/.config/fastly/viceroy: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by /root/.config/fastly/viceroy)

ERROR: error during execution process:
```

**Steps to reproduce:**

Step 1:
```
docker build -t fastly/cli/assemblyscript . -f ./Dockerfile-assemblyscript
docker run -v $PWD:/app -it fastly/cli/assemblyscript
```

Step 2:
```
gh repo clone fastly/compute-starter-kit-assemblyscript-default

or 
git clone git@github.com:fastly/compute-starter-kit-assemblyscript-default.git
```

Step 3:
```
docker run -v $PWD:/app -it fastly/cli/assemblyscript compute init  --from=https://github.com/fastly/compute-starter-kit-assemblyscript-default 

Creating a new Compute@Edge project (using --from to locate package template).

Press ^C at any time to quit.

Name: [app] 
Description: 
Author: 

✓ Initializing...
✓ Fetching package template...
✓ Updating package manifest...
✓ Initializing package...

Initialized package app to:
	/app

To publish the package (build and deploy), run:
	fastly compute publish

To learn about deploying Compute@Edge projects using third-party orchestration tools, visit:
	https://developer.fastly.com/learning/integrations/orchestration/


SUCCESS: Initialized package app

```

Step 4:
```

docker run -v $PWD:/app -it fastly/cli/assemblyscript compute build                            
✓ Initializing...
✓ Verifying package manifest...
✓ Verifying local assemblyscript toolchain...
✓ Building package using assemblyscript toolchain...
✓ Creating package archive...

SUCCESS: Built package 'app' (pkg/app.tar.gz)


```

Step 5:
```
docker run -v $PWD:/app -it fastly/cli/assemblyscript compute serve   
✓ Initializing...
✓ Verifying package manifest...
✓ Verifying local assemblyscript toolchain...
✓ Building package using assemblyscript toolchain...
✓ Creating package archive...

SUCCESS: Built package 'app' (pkg/app.tar.gz)

✓ Checking latest Viceroy release...
✓ Fetching latest Viceroy release...
✓ Running local server...

/root/.config/fastly/viceroy: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by /root/.config/fastly/viceroy)

ERROR: error during execution process:

/root/.config/fastly/viceroy: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by /root/.config/fastly/viceroy).

If you believe this error is the result of a bug, please file an issue: https://github.com/fastly/cli/issues/new?labels=bug&template=bug_report.md



```

**Comment by Integralist** [comment]
Thanks @byteshiva I can confirm that I can replicate this:

```bash
~/Code/go/testing-cli  Tue 18 Jan 2022 17:57:06 GMT
➜  docker run -v $PWD:/app -it fastly/cli/assemblyscript version
Fastly CLI version 1.4.0 (e91af24)
Built with go version go1.17.5 linux/amd64

~/Code/go/testing-cli  Tue 18 Jan 2022 17:57:09 GMT
➜  docker run -v $PWD:/app -it fastly/cli/assemblyscript compute build
✓ Initializing...
✗ Verifying package manifest...

ERROR: error reading package manifest.

Run `fastly compute init` to ensure a correctly configured manifest. See more at https://developer.fastly.com/reference/fastly-toml/

~/Code/go/testing-cli  Tue 18 Jan 2022 17:57:12 GMT
➜  docker run -v $PWD:/app -it fastly/cli/assemblyscript compute init --language assemblyscript

Creating a new Compute@Edge project.

Press ^C at any time to quit.

Name: [app]
Description:
Author:
Starter kit:
[1] Default starter for AssemblyScript
    A basic starter kit that demonstrates routing, simple synthetic responses and
    overriding caching rules.
    https://github.com/fastly/compute-starter-kit-assemblyscript-default
Choose option or paste git URL: [1]

✓ Initializing...
✓ Fetching package template...
✓ Updating package manifest...
✓ Initializing package...

Initialized package app to:
        /app

To publish the package (build and deploy), run:
        fastly compute publish

To learn about deploying Compute@Edge projects using third-party orchestration tools, visit:
        https://developer.fastly.com/learning/integrations/orchestration/


SUCCESS: Initialized package app

~/Code/go/testing-cli is 📦 v1.0.0 via  v16.13.0  Tue 18 Jan 2022 17:57:40 GMT
➜  ls
Octal Permissions Size Date Modified Name
0750  drwxr-x---     - 18 Jan 17:57   assembly
0644  .rw-r--r--    24 18 Jan 17:57   .gitignore
0644  .rw-r--r--   330 18 Jan 17:57   asconfig.json
0644  .rw-r--r--   342 18 Jan 17:57   fastly.toml
0644  .rw-r--r--  7.0k 18 Jan 17:57   package-lock.json
0644  .rw-r--r--   984 18 Jan 17:57   package.json
0644  .rw-r--r--  1.4k 18 Jan 17:57   README.md

~/Code/go/testing-cli is 📦 v1.0.0 via  v16.13.0  Tue 18 Jan 2022 17:57:55 GMT
➜  docker run -v $PWD:/app -it fastly/cli/assemblyscript compute build
✓ Initializing...
✓ Verifying package manifest...
✓ Verifying local assemblyscript toolchain...
✓ Building package using assemblyscript toolchain...
✓ Creating package archive...

SUCCESS: Built package 'app' (pkg/app.tar.gz)

~/Code/go/testing-cli is 📦 v1.0.0 via  v16.13.0  Tue 18 Jan 2022 17:58:24 GMT
➜  docker run -v $PWD:/app -it fastly/cli/assemblyscript compute serve
✓ Initializing...
✓ Verifying package manifest...
✓ Verifying local assemblyscript toolchain...
✓ Building package using assemblyscript toolchain...
✓ Creating package archive...

SUCCESS: Built package 'app' (pkg/app.tar.gz)

✓ Checking latest Viceroy release...
✓ Fetching latest Viceroy release...
✓ Running local server...

/root/.config/fastly/viceroy: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by /root/.config/fastly/viceroy)

ERROR: error during execution process:

/root/.config/fastly/viceroy: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by /root/.config/fastly/viceroy).

If you believe this error is the result of a bug, please file an issue: https://github.com/fastly/cli/issues/new?labels=bug&template=bug_report.md
```

**Comment by Integralist** [comment]
@byteshiva looks like I could resolve this `GLIBC_2.29` error by simply updating the node docker image to the latest version. Let me know if it works for you now.

**Comment by byteshiva** [comment]
It works on `node:17 `

but fails to work for `node:16.13.2` or `node:latest`

```
➜  compute-starter-kit-assemblyscript-default git:(main) ✗ docker run -v $PWD:/app -it --rm fastly/cli/assemblyscript:17 compute build  
✓ Initializing...
✓ Verifying package manifest...
✓ Verifying local assemblyscript toolchain...
✓ Building package using assemblyscript toolchain...
✓ Creating package archive...
```


```
SUCCESS: Built package 'Default-starter-for-AssemblyScript' (pkg/Default-starter-for-AssemblyScript.tar.gz)
➜  compute-starter-kit-assemblyscript-default git:(main) ✗ docker run -v $PWD:/app -it --rm fastly/cli/assemblyscript:17 compute serve  
✓ Initializing...
✓ Verifying package manifest...
✓ Verifying local assemblyscript toolchain...
✓ Building package using assemblyscript toolchain...
✓ Creating package archive...

SUCCESS: Built package 'Default-starter-for-AssemblyScript' (pkg/Default-starter-for-AssemblyScript.tar.gz)

✓ Checking latest Viceroy release...
✓ Fetching latest Viceroy release...
✓ Running local server...

Jan 19 02:34:45.702  WARN no backend definitions found in /app/fastly.toml
Jan 19 02:34:45.702  INFO Listening on http://127.0.0.1:7676

```

**Comment by byteshiva** [comment]
Hi @Integralist 

Can I use the assemblyscript docker instance to build and serve javascript starter kit, 

```
docker run -v $PWD:/app -it --rm fastly/cli/javascript compute init --from=https://github.com/fastly/compute-starter-kit-javascript-default
```

**Comment by Integralist** [comment]
Hi @byteshiva 

re: 

> It works on node:17 but fails to work for node:16.13.2 (LTS) or node:latest

I double checked [DockerHub](https://hub.docker.com/_/node) and it indicates that `latest` is the same image as `17`, so I tested locally using `latest` and it also worked fine for me (which would make sense if they're both built from the same image).

re:

> Can I use the assemblyscript docker instance to build and serve javascript starter kit,

It's _possible_ that the AssemblyScript one will work, but at the time those `Dockerfile`s were created we didn't have the JavaScript platform ready and so it's more likely we'll need to create a separate `Dockerfile` for JavaScript.

**Comment by Integralist** [comment]
Also, just a quick FYI, you don't need to run `compute build` and then `compute serve` as the latter command internally runs the build step. 

If you take a look at the docs (https://developer.fastly.com/reference/cli/compute/serve/) you'll also see you can pass a `--skip-build` flag to `compute serve` if you ever did want to run the command without a build.

**Comment by byteshiva** [comment]
Notes and Observations:

1. The portal fails to render with localhost:7676 but works fine when IP address(192.168.1.2:7676).

![image](https://user-images.githubusercontent.com/10637299/150251748-eca24c78-4e6c-4734-9b83-f78a07a20f31.png)

![image](https://user-images.githubusercontent.com/10637299/150252024-07a87374-b7a2-4d0f-9283-a7a8f69fcffa.png)

![image](https://user-images.githubusercontent.com/10637299/150252315-8349a433-72d8-4102-9f92-e148206e7345.png)



Wasm throws error  "ERROR request{id=2}: WebAssembly trapped: exit with invalid exit status outside of [0..126)
"
```
✗ docker run -v $PWD:/app -it -p 7676:7676 --rm fastly/cli/assemblyscript:17 compute serve  --addr="0.0.0.0:7676" 
✓ Initializing...
✓ Verifying package manifest...
✓ Verifying local assemblyscript toolchain...
✓ Building package using assemblyscript toolchain...
✓ Creating package archive...

SUCCESS: Built package 'Default-starter-for-AssemblyScript' (pkg/Default-starter-for-AssemblyScript.tar.gz)

✓ Checking latest Viceroy release...
✓ Fetching latest Viceroy release...
✓ Running local server...

Jan 20 01:57:42.484  INFO checking if backend 'backend_b' is up
Jan 20 01:57:47.485  WARN backend 'backend_b' on 'https://example.org/' is not up right now
Jan 20 01:57:47.485  INFO checking if backend 'backend_c' is up
Jan 20 01:57:52.486  WARN backend 'backend_c' on 'https://youtube.com/' is not up right now
Jan 20 01:57:52.486  INFO checking if backend 'backend_a' is up
Jan 20 01:57:52.487  WARN backend 'backend_a' on 'http://0.0.0.0/' is not up right now
Jan 20 01:57:52.487  INFO checking if backend 'backend_d' is up
Jan 20 01:57:52.487  WARN backend 'backend_d' on 'http://0.0.0.0:8080/' is not up right now
Jan 20 01:57:52.487  INFO Listening on http://0.0.0.0:7676
Jan 20 01:57:56.399  INFO request{id=0}: handling request GET http://192.168.1.7:7676/
Jan 20 01:57:56.401  INFO request{id=0}: request completed using 65.5 KB of WebAssembly heap
Jan 20 01:58:03.445  INFO request{id=1}: handling request GET http://192.168.1.7:7676/favicon.ico
Jan 20 01:58:03.445  INFO request{id=1}: request completed using 65.5 KB of WebAssembly heap
Jan 20 01:58:07.612  INFO request{id=2}: handling request GET http://localhost:7676/
Jan 20 01:58:07.612  INFO request{id=2}: request completed using 65.5 KB of WebAssembly heap
abort: Failed to construct 'URL': Invalid URL in ~lib/@fastly/as-url/util.ts(15:3)
Jan 20 01:58:07.612 ERROR request{id=2}: WebAssembly trapped: exit with invalid exit status outside of [0..126)
wasm backtrace:
    0:  0x5f1 - <unknown>!<wasm function 16>
    1: 0x47d5 - <unknown>!<wasm function 65>
    2: 0x4f07 - <unknown>!<wasm function 72>
    3: 0x2c78 - <unknown>!<wasm function 55>
```

**Comment by Integralist** [comment]
Thanks @byteshiva for this info. 

The latest errors you're seeing look to be related to [Viceroy](https://github.com/fastly/viceroy) (Viceroy is a local version of Compute@Edge so you can test your code locally). The CLI opens a subprocess to run the `viceroy` binary and those errors you're seeing are subsequently being emitted from Viceroy. 

I'll reach out to the Viceroy team and get them to comment.

**Comment by Integralist** [comment]
Hi @byteshiva 

We have some quick clarifying questions (just to eliminate any clear-cut problems)...

1. It's not clear from the discussion so far if you're using a vanilla starter kit or have added some additional code to it. If the latter, would you be able to share the modifications you've made (such as to the code or package.json etc)?
2. The error suggests that you might not be mounting the source code properly in the container. You're using `-v $PWD:/app ` which mounts the current working directory into the directory `/app` within the container. Therefore, for it to succeed, `$PWD` needs to contain the starter kit source code including a `src` directory. i.e. can we confirm you're running the `docker run` command from the root of an initialised `fastly compute` package?

Thanks.

**Comment by byteshiva** [comment]
I haven't changed anything in the starter assemblyscript kit, 

package.json
```
{
  "name": "compute-starter-kit-assemblyscript-default",
  "version": "1.0.0",
  "description": "Default package starter kit for AssemblyScript based Compute@Edge projects",
  "main": "src/index.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/fastly/compute-starter-kit-assemblyscript-default.git"
  },
  "keywords": [],
  "author": "oss@fastly.com",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/fastly/compute-starter-kit-assemblyscript-default/issues"
  },
  "homepage": "https://github.com/fastly/compute-starter-kit-assemblyscript-default#readme",
  "engines": {
    "node": "^16"
  },
  "devDependencies": {
    "assemblyscript": "^0.19.18"
  },
  "dependencies": {
    "@fastly/as-compute": "^0.4.2"
  },
  "scripts": {
    "asbuild:untouched": "asc assembly/index.ts --target debug",
    "asbuild:optimized": "asc assembly/index.ts --target release",
    "asbuild": "npm run asbuild:untouched && npm run asbuild:optimized"
  }
}

I've only added backend in fastly.toml other than that everything is same.
```

![image](https://user-images.githubusercontent.com/10637299/150325841-5367b85e-8cfa-4e59-b579-5d656d0920fe.png)

**Comment by byteshiva** [comment]
> Thanks @byteshiva for this info.
> 
> The latest errors you're seeing look to be related to [Viceroy](https://github.com/fastly/viceroy) (Viceroy is a local version of Compute@Edge so you can test your code locally). The CLI opens a subprocess to run the `viceroy` binary and those errors you're seeing are subsequently being emitted from Viceroy.
> 
> I'll reach out to the Viceroy team and get them to comment.

Do you want me to raise a separate issue in Viceroy!

**Comment by Integralist** [comment]
@torch2424 Could you maybe review this issue as I might have mis-read the error(s). 

It might be related to something within the AssemblyScript library as per this line...

```
abort: Failed to construct 'URL': Invalid URL in ~lib/@fastly/as-url/util.ts(15:3)
```

Thanks.

**Comment by byteshiva** [comment]
Hi @Integralist 

This issue is again related to Docker,  https://blog.gougousis.net/file-permissions-the-painful-side-of-docker/

Now it works fine. after adding `--user $(id -u)
```
➜  assemblyscript mkdir asm-1
➜  assemblyscript cd asm-1 
➜  asm-1 docker run --user $(id -u) -v $PWD:/app -it --rm fastly/cli/assemblyscript:17 compute init --language assemblyscript 

Creating a new Compute@Edge project.

Press ^C at any time to quit.

Name: [app] 
Description: 
Author: 
Starter kit:
[1] Default starter for AssemblyScript
    A basic starter kit that demonstrates routing, simple synthetic responses and
    overriding caching rules.
    https://github.com/fastly/compute-starter-kit-assemblyscript-default
Choose option or paste git URL: [1] 

✓ Initializing...
- Fetching package template...
✓ Fetching package template...
✓ Updating package manifest...
\ npm WARN EBADENGINE }
✓ Initializing package...

Initialized package app to:
	/app

To publish the package (build and deploy), run:
	fastly compute publish

To learn about deploying Compute@Edge projects using third-party orchestration tools, visit:
	https://developer.fastly.com/learning/integrations/orchestration/


SUCCESS: Initialized package app
➜  asm-1 docker run --user $(id -u) -v $PWD:/app -it -p 7676:7676 --rm fastly/cli/assemblyscript:17 compute serve --addr="0.0.0.0:7676"
✓ Initializing...
✓ Verifying package manifest...
✓ Verifying local assemblyscript toolchain...
✓ Building package using assemblyscript toolchain...
✓ Creating package archive...

SUCCESS: Built package 'app' (pkg/app.tar.gz)

✓ Checking latest Viceroy release...
✓ Fetching latest Viceroy release...
✓ Running local server...

Jan 20 14:56:19.955  WARN no backend definitions found in /app/fastly.toml
Jan 20 14:56:19.955  INFO Listening on http://0.0.0.0:7676
^C
INFO: Local server stopped
➜  asm-1 

```

**Comment by torch2424** [comment]
Hello! First, thank you @byteshiva for all of the super detailed information on the issue! It's super helpful! 😄 And thanks for taking the time to open an issue! 

@Integralist Yes! I think you are 100% correct it's a `@fastly/as-url` bug, related to handling "localhost" urls. And looking through our internal issues and pull requests, we have a PR opened and ready to merge to fix this! So once that is done, I'll go ahead and push out the version, and update this issue 😄

**Comment by torch2424** [comment]
Ah! My apologies! I forgot to update this issue 😢 But! This was published and fixed in: `@fastly/as-url@0.1.2`

https://unpkg.com/browse/@fastly/as-url@0.1.2/CHANGELOG.md

Closing this issue as it should not work with the latest as-url! 😄

**Comment by byteshiva** [comment]
Thanks, @torch2424 

It works with the latest release

![image](https://user-images.githubusercontent.com/10637299/157403285-9fe46dc0-d654-4ea5-813e-23431987ae5f.png)
